### DISCUSSION TITLE
How would you export utils/libs from a nuxt module ?

### DISCUSSION THREAD

**Discussion by hlassiege** [nuxt/nuxt#18603] [Questions]
Hi,

We use nuxt module to share some components, composables and plugins across our apps.
However I would also like to share some common code, interfaces, and types. 
a nuxt module allows me to share nuxt components : composables, plugins, pages etc... but I don't see anything allowing me to export code. I have the feeling it's not possible.

Am I wrong ? How would you build this lib (knowing that this lib would however use some composable) ?

**Comment by dsvgl** [comment]
You can use `addImportsDir` for sharing code dirs. And also share your types (defined in a *.d.ts file) via the nuxt hook `prepare:types`. Eg. like this:

```ts
import { addImportsDir, createResolver, defineNuxtModule } from '@nuxt/kit';

export default defineNuxtModule({
  setup(options, nuxt) {
    const { resolve } = createResolver(import.meta.url);

    const constantsDir = resolve('../constants');
    addImportsDir(constantsDir);

    const typesDir = resolve('../types');
    nuxt.hook('prepare:types', ({ references }) => {
      references.push({ path: `${typesDir}/mytypes.d.ts` });
    });
  },
)};
```

**Comment by dulnan** [comment]
Do you mean helpers/code in general, that can be imported from anywhere within your Nuxt app, incl. nitro (server handlers, etc.?).

I had this issue too where I wanted a method to be available to module users anywhere in their app. I solved it by creating a custom alias that points to a runtime folder from my module:

```typescript
const { resolve } = createResolver(import.meta.url)

nuxt.options.alias['#my-module-utils'] = resolve(
  'runtime/utils',
)
```

I also have this snippet that makes it possible to use the alias/import from a nitro context:

```typescript
nuxt.hook('nitro:config', (nitroConfig) => {
  nitroConfig.externals = defu(
    typeof nitroConfig.externals === 'object' ? nitroConfig.externals : {},
    {
      inline: [resolve('./runtime/utils')],
    },
  )
})
```

`runtime/utils/index.ts` then exports the functions/types/etc. I want to share. These can then be imported as expected:

```typescript
import { getHello } from '#my-module-utils'
```

Here's a quick example showing how it's setup and how utils can be imported in a page component and a custom server route:

https://stackblitz.com/edit/nuxt-starter-atheug?file=pages%2Findex.vue,modules%2FmyModule%2Fruntime%2Futils%2Findex.ts,modules%2FmyModule%2Fmodule.ts&file=server%2Froutes%2Fapi%2Fhello.ts

**Comment by beyer-johannes** [comment]
This is an old common problem since begin of nuxt 3. I hope there will come a more dx friendly way.  Or does we miss something @danielroe ?

Here an old discussion about the same problem: https://github.com/nuxt/nuxt/discussions/16885

**Comment by hlassiege** [comment]
Hi,
thanks for your different answers

I wanted to define an API (types, interfaces etc...) and export it along the composable that use it.
I didn't succeed through the module itself. But I took another solution and create a independent lib just to expose the API. This API is used in the module and in the apps that use the module.
That way it works correctly for me

**Reply by dulnan** [reply]
Okay, so if you only want to export TypeScript types, interfaces, etc. you should be able to do that by exporting them in your module.ts.

What doesn‘t work is exporting actual code directly from the module, types work.
